lib/EPub.php

```diff
@@ -0,0 +1,843 @@
+        $this->meta = $nodes->item(0)->attr('full-path');
+
+        // load metadata
+        if (!$this->zip->FileExists($this->meta)) {
+            throw new Exception('Unable to find ' . $this->meta);
+        }
+
+        $data = $this->zip->FileRead($this->meta);
+        if (!$data) {
+            throw new Exception('Failed to access epub metadata');
+        }
+        $this->xml =  new DOMDocument();
+        $this->xml->registerNodeClass('DOMElement', 'EPubDOMElement');
+        $this->xml->loadXML($data);
+        $this->xml->formatOutput = true;
+        $this->xpath = new EPubDOMXPath($this->xml);
+    }
+
+    public function initSpineComponent()
+    {
+        $spine = $this->xpath->query('//opf:spine')->item(0);
+        $tocid = $spine->getAttribute('toc');
+        $tochref = $this->xpath->query('//opf:manifest/opf:item[@id="' . $tocid . '"]')->item(0)->attr('href');
+        $tocpath = $this->getFullPath($tochref);
+        // read epub toc
+        if (!$this->zip->FileExists($tocpath)) {
+            throw new Exception('Unable to find ' . $tocpath);
+        }
+
+        $data = $this->zip->FileRead($tocpath);
+        $this->toc = new DOMDocument();
+        $this->toc->registerNodeClass('DOMElement', 'EPubDOMElement');
+        $this->toc->loadXML($data);
+        $this->toc_xpath = new EPubDOMXPath($this->toc);
+        $rootNamespace = $this->toc->lookupNamespaceUri($this->toc->namespaceURI);
+        $this->toc_xpath->registerNamespace('x', $rootNamespace);
+    }
+
+    /**
+     * file name getter
+     */
+    public function file()
+    {
+        return $this->file;
+    }
+
+    /**
+     * Close the epub file
+     */
+    public function close()
+    {
+        $this->zip->FileCancelModif($this->meta);
+        // TODO: Add cancelation of cover image
+        $this->zip->Close();
+    }
+
+    /**
+     * Remove iTunes files
+     */
+    public function cleanITunesCrap()
+    {
+        if ($this->zip->FileExists('iTunesMetadata.plist')) {
+            $this->zip->FileReplace('iTunesMetadata.plist', false);
+        }
+        if ($this->zip->FileExists('iTunesArtwork')) {
+            $this->zip->FileReplace('iTunesArtwork', false);
+        }
+    }
+
+    /**
+     * Writes back all meta data changes
+     */
+    public function save()
+    {
+        $this->download();
+        $this->zip->close();
+    }
+
+    /**
+     * Get the updated epub
+     */
+    public function download($file=false)
+    {
+        $this->zip->FileReplace($this->meta, $this->xml->saveXML());
+        // add the cover image
+        if ($this->imagetoadd) {
+            $this->zip->FileReplace($this->coverpath, file_get_contents($this->imagetoadd));
+            $this->imagetoadd='';
+        }
+        if ($file) {
+            $this->zip->Flush(TBSZIP_DOWNLOAD, $file);
+        }
+    }
+
+    /**
+     * Get the components list as an array
+     */
+    public function components()
```

Comment:
All the example I'll use come from `fb.opf` that is inside `test/test.epub`.

First the term component is not a good one, it should be SpineItem or something like that. The method return the book's TOC (called spine in line 202 of `fb.opf`). Instead of simply returning ids (`idref`) if return the full pathname by using the manifest (line 26 of `fb.opf`).

The weird encoding/decoding come from the bad decision before. Instead of using the spine Id, I now have a path that I'll use in a query param (in COPS) -> so I had to encode / decode it.

Using proper idref has been on my Todo list since 2014 :(